lib/charms/postgresql_k8s/v1/postgresql.py

```diff
@@ -202,129 +220,243 @@ def create_access_groups(self) -> None:
-                for data in relation.data.values():
-                    if data.get("database") == database:
-                        relations_accessing_this_database += 1
-            with self._connect_to_database(database=database) as conn, conn.cursor() as curs:
-                curs.execute(
-                    "SELECT schema_name FROM information_schema.schemata WHERE schema_name NOT LIKE 'pg_%' and schema_name <> 'information_schema';"
-                )
-                schemas = [row[0] for row in curs.fetchall()]
-                statements = self._generate_database_privileges_statements(
-                    relations_accessing_this_database, schemas, user
-                )
-                for statement in statements:
-                    curs.execute(statement)
+
+            with self._connect_to_database(database=database) as database_connection, database_connection.cursor() as database_cursor:
+                database_cursor.execute(SQL("REVOKE EXECUTE ON FUNCTION set_user_u(text) FROM {};").format(Identifier(f"{database}_owner")))
+                database_cursor.execute(SQL("REVOKE EXECUTE ON FUNCTION set_user_u(text) FROM {};").format(Identifier(f"{database}_admin")))
+
+                database_cursor.execute(SQL("REVOKE EXECUTE ON FUNCTION set_user(text) FROM {};").format(Identifier(f"{database}_owner")))
+                database_cursor.execute(SQL("REVOKE EXECUTE ON FUNCTION set_user(text) FROM {};").format(Identifier(f"{database}_admin")))
+
+                database_cursor.execute(SQL("REVOKE EXECUTE ON FUNCTION set_user(text, text) FROM {};").format(Identifier(f"{database}_owner")))
+                database_cursor.execute(SQL("REVOKE EXECUTE ON FUNCTION set_user(text, text) FROM {};").format(Identifier(f"{database}_admin")))
+
+                database_cursor.execute(SQL("ALTER DEFAULT PRIVILEGES FOR ROLE {} GRANT SELECT ON TABLES TO {};").format(Identifier(f"{database}_owner"), Identifier(f"{database}_admin")))
+                database_cursor.execute(SQL("ALTER DEFAULT PRIVILEGES FOR ROLE {} GRANT EXECUTE ON FUNCTIONS TO {};").format(Identifier(f"{database}_owner"), Identifier(f"{database}_admin")))
+                database_cursor.execute(SQL("ALTER DEFAULT PRIVILEGES FOR ROLE {} GRANT SELECT ON SEQUENCES TO {};").format(Identifier(f"{database}_owner"), Identifier(f"{database}_admin")))
+
+                database_cursor.execute(SQL("""CREATE OR REPLACE FUNCTION login_hook.login() RETURNS VOID AS $$
+DECLARE
+    ex_state TEXT;
+    ex_message TEXT;
+    ex_detail TEXT;
+    ex_hint TEXT;
+    ex_context TEXT;
+    cur_user TEXT;
+    db_admin_role TEXT;
+    db_name TEXT;
+    db_owner_role TEXT;
+    is_user_admin BOOLEAN;
+BEGIN
+    IF NOT login_hook.is_executing_login_hook()
+    THEN
+        RAISE EXCEPTION 'The login_hook.login() function should only be invoked by the login_hook code';
+    END IF;
+
+    cur_user := (SELECT current_user);
+
+    EXECUTE 'SELECT current_database()' INTO db_name;
```

Comment:
Out of curiosity: does this work in replicas? I had some errors in the past where I tried to use `INTO variable` in the replicas. I'm not sure if it was really related to the unit being a replica or another reason.

Reply:
The PGB auth user also can't be created, since it needs access to user password hashes.